### PR TITLE
Fix the loss scale optimizer for tf.distribute.

### PR DESCRIPTION
**Pull request by qlzh727**
This is trying  to address https://github.com/keras-team/keras/issues/18666

Please note that this fix put backend specific logic into common loss scale optimizer. Let me know if you would like me to create a backend subclass for tensorflow only loss scale optimizer.

### CODE REVIEW

File: integration_tests/tf_distribute_training_test.py [hunk 1]
```diff
@@ -44,7 +44,10 @@ def test_model_fit():
     # Fit from numpy arrays:
     with strategy.scope():
         model.compile(
-            optimizer=optimizers.SGD(learning_rate=0.001, momentum=0.01),
+            # optimizer=optimizers.SGD(learning_rate=0.001, momentum=0.01),
```
Comment (mattdangerw) [inline integration_tests/tf_distribute_training_test.py:47]: should we just delete this commented out line?

File: keras/optimizers/loss_scale_optimizer.py [hunk 2]
```diff
@@ -150,37 +150,87 @@ def apply(self, grads, trainable_variables=None):
...
-            ]
-            self.inner_optimizer.apply(
-                unscaled_grads, trainable_variables=trainable_variables
-            )
+        if backend.backend() == "tensorflow":
+            self._tf_apply(grads, trainable_variables)
+        else:
+            self._common_apply(grads, trainable_variables)
+
+    def handle_finite_grads(self, grads, trainable_variables):
```
Comment (mattdangerw) [inline keras/optimizers/loss_scale_optimizer.py:158]: Why are we pulling these out to toplevels? It is slightly confusing because we also have stateless versions for jax with the same name declared inline.

Probably we should either keep them all inlined, or pull them all out, with a `stateless_` prefix for the stateless ones?

**Review by mattdangerw — APPROVED** [review]
Looks good, just a couple comments.

File: keras/optimizers/loss_scale_optimizer.py (hunk 2 above)
Comment (qlzh727) [reply keras/optimizers/loss_scale_optimizer.py:158]: Thanks for the feedback. The root cause here for tf is that when running with distribution strategy, the cond and the actual apply need to happen in different context (cross-replica vs per-replica). This force us to have those method available at top level so that it can be used by distribution and non-distribution code path.

For the stateless path, I can also lift the handle_x function out, and we probably should add _ to make them private methods.

File: integration_tests/tf_distribute_training_test.py (hunk 1 above)
Comment (qlzh727) [reply integration_tests/tf_distribute_training_test.py:47]: Done. Thanks for catching this.

File: keras/optimizers/loss_scale_optimizer.py (hunk 2 above)
Comment (qlzh727) [reply keras/optimizers/loss_scale_optimizer.py:158]: Done. Updated.

Comment (mattdangerw) [reply keras/optimizers/loss_scale_optimizer.py:158]: don't see the update yet, did you forget to push?

Comment (qlzh727) [reply keras/optimizers/loss_scale_optimizer.py]: ops, just pushed now.

**Review by mattdangerw — APPROVED** [review]
LGTM!
